src/spdx/parser/tagvalue/lexer/tagvalue.py

```diff
@@ -0,0 +1,241 @@
+        "LicenseName": "LICS_NAME",
+        "LicenseCrossReference": "LICS_CRS_REF",
+        "LicenseComment": "LICS_COMMENT",
+        # Snippet
+        "SnippetSPDXID": "SNIPPET_SPDX_ID",
+        "SnippetName": "SNIPPET_NAME",
+        "SnippetComment": "SNIPPET_COMMENT",
+        "SnippetCopyrightText": "SNIPPET_CR_TEXT",
+        "SnippetLicenseComments": "SNIPPET_LICS_COMMENT",
+        "SnippetFromFileSPDXID": "SNIPPET_FILE_SPDXID",
+        "SnippetLicenseConcluded": "SNIPPET_LICS_CONC",
+        "LicenseInfoInSnippet": "SNIPPET_LICS_INFO",
+        "SnippetAttributionText": "SNIPPET_ATTRIBUTION_TEXT",
+        "SnippetByteRange": "SNIPPET_BYTE_RANGE",
+        "SnippetLineRange": "SNIPPET_LINE_RANGE",
+        # Common
+        "NOASSERTION": "NO_ASSERTION",
+        "NONE": "NONE",
+        "SOURCE": "SOURCE",
+        "BINARY": "BINARY",
+        "ARCHIVE": "ARCHIVE",
+        "APPLICATION": "APPLICATION",
+        "AUDIO": "AUDIO",
+        "IMAGE": "IMAGE",
+        "TEXT": "FILETYPE_TEXT",
+        "VIDEO": "VIDEO",
+        "DOCUMENTATION": "DOCUMENTATION",
+        "SPDX": "SPDX",
+        "OTHER": "OTHER",
+        "REVIEW": "REVIEW",
+        "FRAMEWORK": "FRAMEWORK",
+        "LIBRARY": "LIBRARY",
+        "CONTAINER": "CONTAINER",
+        "OPERATING-SYSTEM": "OPERATING_SYSTEM",
+        "DEVICE": "DEVICE",
+        "FIRMWARE": "FIRMWARE",
+        "FILE": "FILE",
+        "INSTALL": "INSTALL"
+    }
+    states = (("text", "exclusive"),)
+
+    tokens = [
+                 "TEXT",
+                 "TOOL_VALUE",
+                 "UNKNOWN_TAG",
+                 "ORG_VALUE",
+                 "PERSON_VALUE",
+                 "DATE",
+                 "LINE",
+                 "CHECKSUM",
+                 "DOC_REF_ID",
+                 "DOC_URI",
+                 "EXT_DOC_REF_CHECKSUM",
+             ] + list(reserved.values())
+
+    def __init__(self):
+        self.lexer = None
+
+    @TOKEN(r":\s*<text>")
+    def t_text(self, t):
+        t.lexer.text_start = t.lexer.lexpos - len("<text>")
+        t.lexer.begin("text")
+
+    @TOKEN(r"</text>\s*")
+    def t_text_end(self, t):
+        t.type = "TEXT"
+        t.value = t.lexer.lexdata[t.lexer.text_start: t.lexer.lexpos]
+        t.lexer.lineno += t.value.count("\n")
+        t.value = t.value.strip()
+        t.lexer.begin("INITIAL")
+        return t
+
+    @TOKEN(r".|\n")
+    def t_text_any(self, t):
+        pass
+
+    def t_text_error(self, t):
+        print("Lexer error in text state")
+
+    @TOKEN(
+        r":\s*(ADLER32|BLAKE2b-256|BLAKE2b-384|BLAKE2b-512|BLAKE3|MD2|MD4|MD5|MD6|SHA1|SHA224|SHA256|SHA384|SHA512|SHA3-256|SHA3-384|SHA3-512):\s*([a-f0-9]*)")
+    def t_CHECKSUM(self, t):
+        t.value = t.value[1:].strip()
+        return t
+
+    @TOKEN(r":\s*DocumentRef-([A-Za-z0-9\+\.\-]+)")
+    def t_DOC_REF_ID(self, t):
+        t.value = t.value[1:].strip()
+        return t
+
+    @TOKEN(r"\s*((ht|f)tps?:\/\/\S*)")
```

Comment:
I see, as I am not that familiar with the whole validation and URI thingy, do you have a suggestion which regex would best match the URI? I think it would need to be quite general on the one hand but shouldn't match the upcoming `Checksum` on the other hand which can be challenging. Another approach would be to not tokenize the `ExternalDocumentRef`, just read it as a line, let the parsing method handle the splitting and the validation would raise a message if the URI doesn't match.

Reply:
I would probably go with the latter: Just read in whatever "URI" is provided, and let validation handle the rest.